src/esp/bindings/GeoBindings.cpp

```diff
@@ -64,6 +66,112 @@ void initGeoBindings(py::module& m) {
+      .def("get_int_voxel", &VoxelWrapper::getVoxel<int>, "index"_a,
+           "grid_name"_a, R"(Gets a voxel at an index in a specified grid.)")
+      .def("get_float_voxel", &VoxelWrapper::getVoxel<float>, "index"_a,
+           "grid_name"_a, R"(Gets a voxel at an index in a specified grid.)")
+      .def("get_vector3_voxel", &VoxelWrapper::getVoxel<Mn::Vector3>, "index"_a,
+           "grid_name"_a, R"(Gets a voxel at an index in a specified grid.)")
+
+      .def("add_bool_grid", &VoxelWrapper::addGrid<bool>, "grid_name"_a,
+           R"(Creates an empty new bool grid.)")
+      .def("add_int_grid", &VoxelWrapper::addGrid<int>, "grid_name"_a,
+           R"(Creates an empty new int grid.)")
+      .def("add_float_grid", &VoxelWrapper::addGrid<float>, "grid_name"_a,
+           R"(Creates an empty new float grid.)")
+      .def("add_vector3_grid", &VoxelWrapper::addGrid<Mn::Vector3>,
+           "grid_name"_a, R"(Creates an empty new vector3 grid.)")
+
+      .def("remove_grid", &VoxelWrapper::removeGrid, "grid_name"_a,
+           R"(Deletes a specified grid and frees its memory.)")
+
+      .def("get_voxel_index_from_global_coords",
+           &VoxelWrapper::getVoxelIndexFromGlobalCoords, "coords"_a,
+           R"(Gets the voxel index located at a coordinate in world space.)")
+      .def(
+          "get_global_coords_from_voxel_index",
+          &VoxelWrapper::getGlobalCoordsFromVoxelIndex, "coords"_a,
+          R"(Gets the world positioning of a particular voxel in a voxelization.)")
+      .def("get_voxel_grid_dimensions", &VoxelWrapper::getVoxelGridDimensions,
+           R"(Returns a numpy array with the dimensions of the voxel grid.)")
+      .def("get_voxel_size", &VoxelWrapper::getVoxelSize,
+           R"(Returns a numpy array with the size of a single voxel.)")
+      // TODO: put voxel grid creation, deletion, and retrieval here
+      .def(
+          "generate_bool_grid_from_int_grid",
+          &VoxelWrapper::generateBoolGridFromIntGrid, "int_grid_name"_a,
+          "bool_grid_name"_a, "start_range"_a, "end_range"_a,
+          R"(Takes in the name of an existing integer grid and produces a boolean grid where each voxel is set to true if the value of that voxel in the integer grid falls within the specified range. The newly created boolean grid is registered under bool_grid_name.)")
+      .def(
+          "generate_bool_grid_from_float_grid",
+          &VoxelWrapper::generateBoolGridFromFloatGrid, "float_grid_name"_a,
+          "bool_grid_name"_a, "start_range"_a, "end_range"_a,
+          R"(Takes in the name of an existing float grid and produces a boolean grid where each voxel is set to true if the value of that voxel in the float grid falls within the specified range. The newly created boolean grid is registered under bool_grid_name.)")
+      // TODO: put voxel set filling functions here (need to figure out how to
+      // pass in function ptr as arg)
```

Comment:
A `std::function` works, but a plain function pointer won't (to make it work, pybind attaches a lot of state to each function).

However, an API in this form will be very slow due to all the jumps between C++ and Python in the callback, so it'll make more sense to do that processing directly in python over the returned view (...once i implement that).